[c10d][flight recorder] store a copy of string in entry

shuqiangzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119837

Summary:
Previously, we just store the char pointer in entry, the string is a
temp object and will be destructed when we want to dump/access it.

A quick fix is to store a copy of the string, but without changing the
upstream char*.

An alternative is to change every profilingTitle into std:string, this
however would needs comprehensive overhall of the code up to the
c10d::work layer above workNCCL and RecordFunction etc.

We chose the first option for this change

Resolve #119808

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

shuqiangzhang committed:
[10d][flight recorder] store a copy of string in entry
Summary:
Previously, we just store the char pointer in entry, the string is a
temp object and will be destructed when we want to dump/access it.

A quick fix is to store a copy of the string, but without changing the
upstream char*.

An alternative is to change every profilingTitle into std:string, this
however would needs comprehensive overhall of the code up to the
c10d::work layer above workNCCL and RecordFunction etc.

We chose the first option for this change

[ghstack-poisoned]

wconstab commented:
well, there are 2 potential things here
1) technically (pedantically, perhaps) this fix is still not totally safe, right?  

void foo(const char* blah) {
   // creates a copy of the contents of blah
  std::string unsafe(blah);
}

// destroys the temporary std::string before even calling foo, so technically the c_str() pointer is invalid by the time foo uses it to create a copy?
foo(c10::string("asdf").c_str());

2) it's a pessimization to store a std::string in Event instead of a ptr.  I'm not really sure if it matters in practice though.  Any idea how expensive it is to construct? (does it heap allocate in addition to copying?)

zdevito commented:
1. I always forget the precise rules, but I think that rvalue string stays alive until the next synchronization point, so its data is valid during the foo call. Still will construct 2 strings in the circumstance, so using a by-value std::string is a better API to have anyway.

zdevito commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -430,7 +430,7 @@ struct NCCLTraceBuffer {
       size_t pg_id,
       size_t seq_id,
-      const char* profiling_name,
+      std::string profiling_name,
```
We should also remove the `.c_str()` call on the other end to avoid creating two strings.

shuqiangzhang replied:
mentioned in the summary too, turns out layers of existing apis (Work, WorkNCCL, initWork) take c_str, so we need to do this kind of string->c_str->string conversions, not ideal for sure

zdevito reviewed (approved):
I think it is find to use std::string. It's probably the case that strings < 20 characters don't even allocate memory because of short string optimization anyway.

shuqiangzhang commented:
for 1), I asked chatgpt (^_^) before implementation to verify that the temporary object will remain valid for the duration of the function call of foo, so hope it is correct

for 2), if I understand correctly, the entries should be created in data segment (global variables), not heap, not stack.

shuqiangzhang committed:
Update on "[10d][flight recorder] store a copy of string in entry"


Summary:
Previously, we just store the char pointer in entry, the string is a
temp object and will be destructed when we want to dump/access it.

A quick fix is to store a copy of the string, but without changing the
upstream char*.

An alternative is to change every profilingTitle into std:string, this
however would needs comprehensive overhall of the code up to the
c10d::work layer above workNCCL and RecordFunction etc.

We chose the first option for this change

Resolve #119808

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang commented:
@pytorchbot merge

wconstab reviewed (approved):
ok, just have to fix the lint and good to go! thanks @shuqiangzhang

shuqiangzhang committed:
Update on "[c10d][flight recorder] store a copy of string in entry"


Summary:
Previously, we just store the char pointer in entry, the string is a
temp object and will be destructed when we want to dump/access it.

A quick fix is to store a copy of the string, but without changing the
upstream char*.

An alternative is to change every profilingTitle into std:string, this
however would needs comprehensive overhall of the code up to the
c10d::work layer above workNCCL and RecordFunction etc.

We chose the first option for this change

Resolve #119808

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang commented:
@pytorchbot merge
